scapy/contrib/isotp.py

```diff
@@ -57,100 +52,1171 @@ def answers(self, other):
+        :param rx_separation_time_min: minimum desired separation time sent in
+                                       Flow Control ISOTP frames
+        :param basecls: base class of the packets emitted by this socket
+        """
+
+        self.src = sid
+        self.dst = did
+        self.exsrc = extended_addr
+        self.exdst = extended_rx_addr
+        self.timeout = timeout
+
+        def can_send(load):
+            can_socket.send(CAN(identifier=sid, data=load))
+
+        self.impl = ISOTPSocketImplementation(
+            can_send,
+            extended_addr=extended_addr,
+            extended_rx_addr=extended_rx_addr,
+            rx_block_size=rx_block_size,
+            rx_separation_time_min=rx_separation_time_min
+        )
+
+        self.can_socket = can_socket
+        self.rx_thread = CANReceiverThread(can_socket, self.impl.on_recv, did)
+        self.rx_thread.start()
+
+        if basecls is None:
+            warning('Provide a basecls ')
+        self.basecls = basecls
+
+    def __del__(self):
+        """Close the socket and stop the receiving thread"""
+        self.can_socket.close()
+        self.rx_thread.stop()
+
+    def __enter__(self):
+        """Using the with statement is useful in unit tests executed on an
+        interactive shell, to make sure the receiver thread is stopped"""
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        """Close the socket and stop the receiving thread"""
+        self.rx_thread.stop()
+
+    def close(self):
+        """Close the socket and stop the receiving thread"""
+        self.can_socket.close()
+        self.rx_thread.stop()
+        SuperSocket.close(self)
+
+    def begin_send(self, p):
+        """Begin the transmission of message p. This method returns after
+        sending the first frame. If multiple frames are necessary to send the
+        message, this socket will unable to send other messages until either
+        the transmission of this frame succeeds or it fails."""
+        if hasattr(p, "sent_time"):
+            p.sent_time = time.time()
+
+        return self.impl.begin_send(bytes(p))
+
+    def send(self, p):
+        """Send ISOTP message p, blocking until either the transmission is
+        successful or it fails.
+        Returns None if the transmission fails, or the number of bytes sent
+        if it succeeds."""
+        if hasattr(p, "sent_time"):
+            p.sent_time = time.time()
+
+        return self.impl.send(bytes(p))
+
+    def recv_raw(self, x=0xffff):
+        """Receive a complete ISOTP message, blocking until a message is
+        received or the specified timeout is reached.
+        If self.timeout is 0, then this function doesn't block and returns the
+        first frame in the receive buffer or None if there isn't any."""
+        return self.basecls, self.impl.recv(self.timeout), time.time()
+
+    def recv(self, x=0xffff):
+        msg = SuperSocket.recv(self, x)
```

Comment:
Is this the correct way, to call a super class? Would this approach be better? 
```
class Foo(Bar):
    def baz(self, arg):
        return super(Foo, self).baz(arg)
```

Reply:
I tried your example with the following code but it does not work on python 2:
```
class Bar:
    def baz(self, arg):
        pass

class Foo(Bar):
    def baz(self, arg):
        return super(Foo, self).baz(arg)

Foo().baz("a")
```